### PR TITLE
Unit test for TURN connection send data lock

### PR DESCRIPTION
**Pull request by zhiyua-git**
*Issue #, if available:*
KinesisVideo - 5340

*Description of changes:*
Adding unit test for the lock added for https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/75e47eb09d2ef491f962a8d04dafd5a0882cd748.
Thread sanitizer would fail if the above commit is reverted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tst/TurnConnectionFunctionalityTest.cpp [hunk 1]
```diff
@@ -263,6 +263,92 @@ namespace com { namespace amazonaws { namespace kinesis { namespace video { name
...
     }
 
+    struct TurnCustomData {
+        PTurnConnection pTurnConnection;
+        PBYTE pBuf;
+        UINT32 bufLen;
+        PKvsIpAddress pKvsIpAddress;
+    };
+
+    PVOID testTurnSendDataRoutine(PVOID arg) {
```
Comment (Sean-Der) [inline tst/TurnConnectionFunctionalityTest.cpp]: Would you mind making this a lambda/keeping it internal to the test if possible?

Just keeps the code a little cleaner for the reader/don't have to worry about others re-using incorrectly.

File: tst/TurnConnectionFunctionalityTest.cpp [hunk 2]
```diff
@@ -263,6 +263,92 @@ namespace com { namespace amazonaws { namespace kinesis { namespace video { name
...
+            if (pTurnConnection->state == TURN_STATE_READY) {
+                turnReady = TRUE;
+            }
+            MUTEX_UNLOCK(pTurnConnection->lock);
+        }
+
+        EXPECT_TRUE(turnReady == TRUE);
+
+        for (i = 0; i < reqCount; i++) {
+            buf[i] = (PBYTE) MEMALLOC(bufLen);
```
Comment (Sean-Der) [inline tst/TurnConnectionFunctionalityTest.cpp]: If possible I would use a `unique_ptr` here, better then having to free manually.

File: tst/TurnConnectionFunctionalityTest.cpp [hunk 3]
```diff
@@ -263,6 +263,92 @@ namespace com { namespace amazonaws { namespace kinesis { namespace video { name
...
+            for (j = 0; j < bufLen; j++) {
+                buf[i][j] = i;
+            }
+            args[i].pBuf = buf[i];
+            args[i].bufLen = bufLen;
+            args[i].pKvsIpAddress = &turnPeerAddr;
+            args[i].pTurnConnection = pTurnConnection;
+            EXPECT_EQ(STATUS_SUCCESS, THREAD_CREATE(&threads[i], testTurnSendDataRoutine, (PVOID) &args[i]));
```
Comment (Sean-Der) [inline tst/TurnConnectionFunctionalityTest.cpp]: I would use `std::thread`, you can capture multiple things and don't have to worry about putting in a struct. The PeerConnectionFunctionality tests do that.

File: tst/TurnConnectionFunctionalityTest.cpp (hunk 1 above)
Comment (zhiyua-git) [reply tst/TurnConnectionFunctionalityTest.cpp]: It is only defined in this cpp and not in any .h file, I suppose that would not be reused and just internal to this test file.

File: tst/TurnConnectionFunctionalityTest.cpp [hunk 4]
```diff
@@ -263,6 +263,73 @@ namespace com { namespace amazonaws { namespace kinesis { namespace video { name
...
+        BOOL turnReady = FALSE;
+        KvsIpAddress turnPeerAddr;
+        UINT32 bufLen = 5;
+        UINT32 reqCount = 10;
+        PBYTE *buf;
+        std::thread *threads;
+        UINT32 i, j;
+
+        buf = (PBYTE*) MEMALLOC(SIZEOF(PBYTE) * reqCount);
+        threads = (std::thread*) MEMALLOC(SIZEOF(std::thread) * reqCount);
```
Comment (MushMal) [inline tst/TurnConnectionFunctionalityTest.cpp]: Use stack based arrays - these are tiny

File: tst/TurnConnectionFunctionalityTest.cpp [hunk 5]
```diff
@@ -263,6 +263,73 @@ namespace com { namespace amazonaws { namespace kinesis { namespace video { name
...
+        EXPECT_EQ(STATUS_SUCCESS, turnConnectionAddPeer(pTurnConnection, &turnPeerAddr));
+        EXPECT_EQ(STATUS_SUCCESS, turnConnectionStart(pTurnConnection));
+
+        // wait until channel is created
+        while(!turnReady) {
+            MUTEX_LOCK(pTurnConnection->lock);
+            if (pTurnConnection->state == TURN_STATE_READY) {
+                turnReady = TRUE;
+            }
+            MUTEX_UNLOCK(pTurnConnection->lock);
```
Comment (MushMal) [inline tst/TurnConnectionFunctionalityTest.cpp]: This is a tight loop. At least add some thread sleeps

**Pull request merged**